elephant/spectral.py

```diff
@@ -523,6 +529,543 @@ def multitaper_psd(signal, n_segments=1, len_segment=None,
+        given.
+        Default: 1.
+    len_segment : int, optional
+        Length of segments. This parameter is ignored if `frequency_resolution`
+        is given. If None, it will be determined from other parameters.
+        Default: None.
+    frequency_resolution : pq.Quantity or float, optional
+        Desired frequency resolution of the obtained PSD estimate in terms of
+        the interval between adjacent frequency bins. When given as a `float`,
+        it is taken as frequency in Hz.
+        If None, it will be determined from other parameters.
+        Default: None.
+    overlap : float, optional
+        Overlap between segments represented as a float number between 0 (no
+        overlap) and 1 (complete overlap).
+        Default: 0.5 (half-overlapped).
+    fs : float, optional
+        Specifies the sampling frequency of the input time series
+        Default: 1.0.
+    nw : float, optional
+        Time-halfbandwidth product. This parameter can be used to determine the
+        number of tapers following the equation:
+            num_tapers = 2*nw - 1
+        It can be determined by multplying the duration of the signal with the
+        desired half-peak resolution frequency:
+            n_samples/fs * peak_resolution/2.
+        Default: 4.0.
+    num_tapers : int, optional
+        Number of tapers used in step 2 (see above) to obtain estimate of PSD.
+        By default, [2*nw] - 1 is chosen.
+        Default: None.
+    peak_resolution : pq.Quantity float, optional
+        Quantity in Hz determining the number of tapers used for analysis.
+        Fine peak resolution --> low numerical value --> low number of tapers
+        High peak resolution --> high numerical value --> high number of tapers
+        When given as a `float`, it is taken as frequency in Hz.
+        Default: None.
+    return_onesided : bool, optional
+        If True, return a one-sided spectrum for real data.
+        If False return a two-sided spectrum.
+        Default: True.
+
+    Returns
+    -------
+    freqs : np.ndarray
+        Frequencies associated with power estimate in `psd`
+    cross_spec : np.ndarray
+        Estimate of the cross spectrum of time series in `signal`
+
+    """
+    # When the input is AnalogSignal, fetch the underlying numpy array and swap
+    # axes from (n_samples, n_channels) to (n_channels, n_samples)
+    data = np.asarray(signals)
+    if isinstance(signals, neo.AnalogSignal):
+        data = np.moveaxis(data, 0, 1)
+        fs = signals.sampling_rate
+
+    # Initialize argument dictionary for multitaper_cross_spectrum called on
+    # segments
+    cross_spec_params_dict = {
+        'fs': fs,
+        'nw': nw,
+        'num_tapers': num_tapers,
+        'peak_resolution': peak_resolution,
+        'return_onesided': return_onesided,
+        'attach_units': False}
+
+    # Apply segmentation
+    freqs, cross_spec_estimate = _segmented_apply_func(
```

Comment:
Renaming the private function to a more telling name of these combined steps will facilitate understanding the code. Something like `_segmented_apply_func_and_average`. By a segmented apply function, one could also think of a function that splits the data, applies the function to each segment, and returns arrays with each segment estimate.

Reply:
This was done to leave the possibility of using this function (with a few edits) to calculate the spectrogram. The spectrogram is not the focus of this PR and, thus, is not included here.